stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt

```diff
@@ -467,50 +440,58 @@ internal constructor(
         )
     }
 
-    private fun setAnonymousUser(listener: InitConnectionListener? = null) {
-        if (userStateService.state is UserState.NotSet) {
+    private suspend fun setAnonymousUser(): Result<ConnectionData> {
+        return if (userStateService.state is UserState.NotSet) {
             socketStateService.onConnectionRequested()
             userStateService.onSetAnonymous()
-            connectionListener = object : InitConnectionListener() {
-                override fun onSuccess(data: ConnectionData) {
-                    initializationCoordinator.userConnected(data.user)
-                    listener?.onSuccess(data)
-                }
-
-                override fun onError(error: ChatError) {
-                    listener?.onError(error)
-                }
-            }
             config.isAnonymous = true
             warmUp()
             socket.connectAnonymously()
+            waitConnection.first().also { result ->
+                if (result.isSuccess) {
+                    initializationCoordinator.userConnected(result.data().user)
+                }
+            }
         } else {
             logger.logE("Failed to connect user. Please check you don't have connected user already")
-            listener?.onError(ChatError("User cannot be set until previous one is disconnected."))
+            Result.error(ChatError("User cannot be set until previous one is disconnected."))
         }
     }
 
     @CheckResult
     public fun connectAnonymousUser(): Call<ConnectionData> {
-        return createInitListenerCall { initListener -> setAnonymousUser(initListener) }
+        return CoroutineCall(scope) {
+            logger.logD("[connectAnonymousUser] no args")
+            setAnonymousUser().also { result ->
+                logger.logV(
+                    "[connectAnonymousUser] completed: ${
+                    result.stringify { "ConnectionData(connectionId=${it.connectionId})" }
+                    }"
+                )
+            }
+        }
     }
 
-    private fun setGuestUser(userId: String, username: String, listener: InitConnectionListener? = null) {
-        getGuestToken(userId, username).enqueue { result ->
+    @CheckResult
+    public fun connectGuestUser(userId: String, username: String): Call<ConnectionData> {
+        return CoroutineCall(scope) {
+            logger.logD("[connectGuestUser] userId: '$userId', username: '$username'")
+            val result = getGuestToken(userId, username).await()
             if (result.isSuccess) {
```

Comment:
You can use mapSuspend to avoid the if/else. 

```
getGuestToken(userId, username).await()
                .mapSuspend { guestUser ->
                    setUser(guestUser.user, ConstantTokenProvider(guestUser.token))
                }
                .data()
                .also { finalResult ->
                    logger.logV(
                        "[connectAnonymousUser] completed: ${
                            finalResult.stringify { "ConnectionData(connectionId=${it.connectionId})" }
                        }"
                    )
                }
```

Not mandatory, do if you see an improvement.